### DISCUSSION TITLE
NFS Ganesha Support

### DISCUSSION THREAD

**Discussion by anon314159** [moosefs/moosefs#611] [Q&A]
Hello MooseFS community,

Does MooseFS support integration with NFS-ganesha?  My reasoning is because attempts to re-export MFS fuse mounts via SMB results in horrendous reads speeds due the mfsmount process wasting CPU cycles calculating CRC values for any uncached files.

**Comment by chogata** [comment]
You have asked about Ganesha already once and gotten a response, in this thread: https://github.com/moosefs/moosefs/issues/594

Also, I explained to you that the crc calculations you have shown are not done by MooseFS, as they were C++ code and MooseFS is pure C. Did you test MooseFS 4, as suggested?

**Reply by anon314159** [reply]
Thanks for the reply.

I just completed testing with NFS-ganesha this morning with the latest version of MooseFS and determined the problem is replicatable and reproducible. Whether its LizardFS, SaunaFS, or MooseFS.  So this is no longer a Samba issue, it's either related to FUSE or the mfsmount process, since both must use the same process to interact with the file system.  A simple stack trace reveals multiple futex wait states when invoking the mycrc32_init() function located within the source file below:

https://github.com/moosefs/moosefs/blob/master/mfsclient/mfsmount.c

It seems like fuse or MFS really does like any form of protocol indirection.  While I am aware Samba isn't fully supported, indirectiom using NFS via Ganesha (VFS) seemed like a valid alternative but alas, it also has the same drop off with uncached read performance.

Here are the steps to reproduce the problem 

1.) Mount MooseFS export on the target proxy server
2). Export this above mount point via NFS-ganesha using VFS
3.) Mount the exported NFS share on proxy server or any client
4.) Attempt to read any file that hasn't been cached yet from NFS mount (cp, dd, rsync, etc.)
5.) Compare the performance delta between native FUSE Mount and NFS re-export.

**Comment by chogata** [comment]
"multiple futex wait states when invoking the mycrc32_init() function" - well, mycrc32_init() is called only once, when a process starts (as the name init would suggest), so this is really strange

Could you do an experiment? Compare, on your hardware and your configuration, the read speed of a special file .random, localized in the MooseFS mountpoint. (E.g. if you mounted MooseFS in /mnt/mfs directory, the file to read would be /mnt/mfs/.random). Read it using (for example) dd command, directly via FUSE and through a ganesha NFS export. What is the difference?

**Reply by anon314159** [reply]
Thanks for the quick reply.

Describe the problem you observed.
Excessively high CPU time spend recalculating uncached chunks affecting the mfsmount process. This has the net byproduct of slowing down read performance when used in conjunction with NFS-Ganesha (VFS) reexports. Benchmarks reveal a near 10-fold decrease in sequential and random reads when compared to the native FUSE client. Supposedly the developers support reexporting MFS to NFS via Ganesha (VFS) but in practice the read speeds are generally terrible for any real-world production use.

Example of Sequential Read Tests:
Cluster-A (Average 10 Runs, 8GB Test File, MooseFS FUSE):
echo 3 > /proc/sys/vm/drop_caches
mount -t moosefs areaxx-xx-001p: /mnt/mfs -o mfsmaster=areaxx-xx-001p.fake.domain.local,mfsport=9421,mfssubfolder=/mfs_export
if=/mnt/mfs/.random of=/dev/null bs=1M count=8192 status=progress
8589934592 bytes (8.6 GiB, 8.0 GiB) copied, 10.3241 s, 830 MB/s

Cluster-A (Average 10 Runs, 8GB Test File, MooseFS FUSE NFS reexport):
mount -t nfs nfs-server:/mfs-test /mnt/nfs/ -o proto=tcp,sec=sys,vers=4
echo 3 > /proc/sys/vm/drop_caches
if=/mnt/nfs/.random of=/dev/null bs=1M count=8192 status=progress
8589934592 bytes (8.6 GiB, 8.0 GiB) copied, 103.2541 s, 83.2 MB/s

Cluster-A (Average 10 Runs, 8GB Test File, MooseFS FUSE SMB reexport):
mount -t smb /areaxx-xx-001p/smb /mnt/smb/ -o domain=fake.domain.local,username=fakeuser
echo 3 > /proc/sys/vm/drop_caches
if=if=/mnt/smb/.random of=/dev/null bs=1M count=8192 status=progress
8589934592 bytes (8.6 GiB, 8.0 GiB) copied, 135.2881 s, 63.4 MB/s

Observation reveals all cached reads bypass the mfsmount process entirely and instead use the operating systems kernels page cache.  Whereas uncached reads operations through MooseFS's FUSE client offered the expected performance but the same can't be said for NFS-Ganesha or SMB.  The phenomenon is reproducible across several clusters regardless of the hardware profiles, albeit my all flash nvme cluster performs the fastest. Cluster-A is built around traditional 7200 RPM SAS drives.

**Reply by chogata** [reply]
This test proves what I was afraid of this whole time - the solution to the problem lies outside of our current scope. The ".random" is a randomly generated file, very fast, and it is wholly generated by the mount process - no communication with master or chunkservers occurs, there is no cache'ing of data, nothing. The speed depends on your CPU, not your drives :)
This means that for whatever reason a FUSE filesystem works faster, when used directly and slower, when re-exported via NFS or SMB. Most probably, this is a FUSE issue - you could try to open an issue with them and present this.
